src/software/new_geom/vector.h

```diff
@@ -0,0 +1,521 @@
+#pragma once
+
+#include <boost/geometry.hpp>
+#include <cmath>
+#include <iostream>
+
+#include "software/new_geom/angle.h"
+
+/**
+ * A vector in 2D space.
+ */
+class Vector final
+{
+   public:
+    /**
+     * Creates a zero-vector.
+     */
+    explicit constexpr Vector();
+
+    /**
+     * Creates a unit-magnitude Vector from an angle.
+     *
+     * @param angle the angle
+     *
+     * @return Vector the Vector
+     */
+    static Vector createFromAngle(const Angle &angle);
+
+    /**
+     * Creates a Vector with arbitrary x and y values.
+     *
+     * @param x the <var>x</var> value of the Vector
+     * @param y the <var>y</var> value of the Vector
+     */
+    constexpr Vector(double x, double y);
+
+    /**
+     * Creates a new Vector that is a copy of the given Vector
+     *
+     * @param the Vector to duplicate
+     */
+    constexpr Vector(const Vector &v);
+
+    /**
+     * Returns the magnitude in the x-coordinate of this Vector
+     *
+     * @return the magnitude in the x-coordinate of this Vector
+     */
+    constexpr double x() const;
+
+    /**
+     * Returns the magnitude in the y-coordinate of this Vector
+     *
+     * @return the magnitude in the y-coordinate of this Vector
+     */
+    constexpr double y() const;
+
+    /**
+     * Sets the magnitudes of this vector to the new magnitudes
+     *
+     * @param x the new magnitude in the x-coordinate
+     * @param y the new magnitude in the y-coordinate
+     */
+    void set(double x, double y);
+
+    /**
+     * Sets the magnitude in the x-coordinate of this vector
+     *
+     * @param x the new x-coordinate magnitude
+     */
+    void setX(double x);
+
+    /**
+     * Sets the magnitude in the y-coordinate of this vector
+     *
+     * @param y the new y-coordinate magnitude
+     */
+    void setY(double y);
+
+    /**
+     * Returns the square of the length of the Vector
+     *
+     * @return the square of the length of the Vector
+     */
+    constexpr double lensq() const;
```

Comment:
Do we want to take this opportunity to un-shorten these function names so things are more readable? IIRC we thought of this a while ago but it wasn't needed at the time. @garethellis0 ?

Reply:
yes, absolutely. This should be `lengthSquared`

Reply:
Same sort of thing goes if there are other similar names.